packages/@sanity/form-builder/src/inputs/ObjectInput/ObjectInput.tsx

```diff
@@ -74,8 +79,24 @@ export const ObjectInput = memo(
     } = props
 
     const handleFieldChange = React.useCallback(
-      (fieldEvent: PatchEvent, field) => {
+      (fieldEvent: PatchEvent, field: ObjectField) => {
         let event = fieldEvent.prefixAll(field.name)
+
+        // check if the result of the incoming patches will result in an empty object.
+        // first apply all the patches to the current value to get the result.
+        const result = applyAll(
+          value || {},
+          // unset the `_type` when computing the result since its auto-populated
+          event.append(unset(['_type'])).patches
+        )
+
+        // if the result has no keys left in it
+        if (isRecord(result) && Object.keys(result).length === 0) {
```

Comment:
Can we add a note about how `_key` for array objects fit into this? Currently, objects in arrays are retained because they have a `_key` property. This is how it has to work, but it seems rather coincidental from reading this code.